src/messaging/MessagingGlobals.h

```diff
@@ -0,0 +1,34 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```

Comment:
nit: 2021

Reply:
Currently,  InetLayer and SystemLayer are defined as global instance, but not ExchangeMgr, later we will have SecurityMgr and FabricState, etc ...  All of them are global singletons and must be declared only once in each app.

I originally want to define this ExchangeMgr in platform/Globals.cpp, but I found this will cause DeviceLayer depend on MessagingLayer. So I borrowed the pattern from OpenWeave, add a Globals.h/Globals.cpp in message layer to declare those singletons from MessagingLayer.

Reply:
I am thinking of the reasoning to define globals: it is to easily access them from various locations, which in my mind is very strong coupling against a global state.

System and Inet I imagine are equivalent to O/S capabilities generally available, so coupling does not seem as bad (although realistically, we should use 'transports' instead of raw system/inet calls). For my own preference I would rather not have these globals including exchange/security/fabric etc. Not sure if possible since some things are obviously global (persistent storage comes in mind, probably so is the address resolution/dns-sd stuff).

Declaring a global seems like a last resort rather than a default way to do things. For my part I am giving a "-1" to this approach.

Reply:
Another reason to define those global instance is to allow app developer focus on using the external APIs of CHIP stack instead of understanding the internal implementation of the stack. Otherwise, the app developer needs to declare all needed global singletons but not really using them in their application logic.

Reply:
globals are 

> I am weary of allowing a global header since I fear coupling: this makes it much easier to include the header and just use `chip::Messaging::ExchangeMgr` instead of having it an explicit dependency.
> 
> I would love feedback from others. My default throughout was generally to never have globals, even if it means the app will need to declare a bunch of them.

I have the same wariness.  Globals are hard to fix after the fact.

Reply:
> Another reason to define those global instance is to allow app developer focus on using the external APIs of CHIP stack instead of understanding the internal implementation of the stack. Otherwise, the app developer needs to declare all needed global singletons but not really using them in their application logic.

declaring all the needed managers can be facilitated by a wrapper class.  I'd expect the set of managers to differ among clients of the CHIP API.  E.g. controllers need a different set than accessories.

Reply:
+1 to minimizing globals. They hide dependencies between objects which can lead to a number of harms such as state leaks between tests, thread safety issues, etc. I think we should avoid them unless they are proven necessary to hit code size targets. Ideally it would be possible to create two completely separate CHIP stacks in the same process that listen on distinct IP addresses (from that perspective, persistent storage could not be global, either). We are pretty far from this currently.

Reply:
Agree with @mspang, we probably need multiple separate CHIP stacks when implementing a CHIP controller which support multiple admins (or ecosystems).

Reply:
Currently, we have the following singleton instances in CHIP to make it work: InetLayer/SystemLayer/ExchangeManager/SecureSessionMgr, and FabricState/SecurityManager later.

Like Weave, those singleton instances should be declared in data segment instead of stack to fit into deeply embedded devices. I don't think we should leave it to third part developers to define them since they are always needed cross all apps and third part developers might define them in an inappropriate place or declare multiple instance of a singleton.

We will have accessory API later, maybe we should wrap all the needed managers behind the unified accessory API after it is ready.

Reply:
I would work towards removing the Exchangemanager and SecureSesionmanager singletons if they are within the library.

We can provide helpers (maybe a class that contains all required members and calls init in a good order).